### PR TITLE
Psychohistory

### PR DESCRIPTION
**Pull request by eggrobin**

### CODE REVIEW

File: ksp_plugin/interface_vessel.cpp [hunk 1]
```diff
@@ -53,6 +53,8 @@ void principia__VesselIncrementIntrinsicForce(
     XYZ const intrinsic_force_in_kilonewtons) {
   journal::Method<journal::VesselIncrementIntrinsicForce> m(
       {plugin, vessel_guid, intrinsic_force_in_kilonewtons});
+  // TODO(egg): this is profoundly incorrect! the vector is given in World, not
```
Comment (pleroy) [inline ksp_plugin/interface_vessel.cpp]: Should be TODO(phl), I think.

File: ksp_plugin/pile_up.cpp [hunk 2]
```diff
@@ -1,19 +1,115 @@
...
+using geometry::Position;
+using physics::DegreesOfFreedom;
+using physics::RigidMotion;
+using physics::RigidTransformation;
+
 PileUp::PileUp(std::list<not_null<Vessel*>>&& vessels)
-    : vessels_(std::move(vessels)) {}
+    : vessels_(std::move(vessels)) {
+  BarycentreCalculator<DegreesOfFreedom<Barycentric>, Mass> barycentre;
+  for (not_null<Vessel*> vessel : vessels_) {
```
Comment (pleroy) [inline ksp_plugin/pile_up.cpp]: `const`

File: ksp_plugin/vessel.cpp [hunk 3]
```diff
@@ -324,9 +328,33 @@ not_null<std::unique_ptr<Vessel>> Vessel::ReadFromMessage(
...
+  vessel->psychohistory_.Append(vessel->history_->last().time(),
+                                vessel->history_->last().degrees_of_freedom());
+  vessel->psychohistory_is_history_ = true;
   return std::move(vessel);
 }
 
+void Vessel::AppendToPsychohistory(
+    Instant const& time,
+    DegreesOfFreedom<Barycentric> const& degrees_of_freedom,
+    bool historical) {
```
Comment (pleroy) [inline ksp_plugin/vessel.cpp]: This is called `authoritative` in the `.h` file.

File: ksp_plugin/vessel.hpp [hunk 4]
```diff
@@ -163,6 +163,14 @@ class Vessel {
       not_null<Ephemeris<Barycentric>*> ephemeris,
       not_null<Celestial const*> parent);
 
+  void AppendToPsychohistory(
+      Instant const& time,
+      DegreesOfFreedom<Barycentric> const& degrees_of_freedom,
+      bool authoritative);
+
+  DiscreteTrajectory<Barycentric> const& psychohistory() const;
+  bool psychohistory_is_history() const;
```
Comment (pleroy) [inline ksp_plugin/vessel.hpp]: This is a rather confusing identifier.  I would prefer `psychohistory_is_authoritative`.

File: ksp_plugin/vessel.cpp [hunk 5]
```diff
@@ -116,6 +116,9 @@ void Vessel::CreateHistoryAndForkProlongation(
   CHECK(!is_initialized());
   history_ = std::make_unique<DiscreteTrajectory<Barycentric>>();
   history_->Append(time, degrees_of_freedom);
+  // TODO(egg): proper initialization.
+  psychohistory_.Append(time, degrees_of_freedom);
+  psychohistory_is_history_ = true;
```
Comment (pleroy) [inline ksp_plugin/vessel.cpp]: I don't understand this.  Shouldn't it be `false` for compatibility?  Same question at line 229.

File: ksp_plugin/pile_up.cpp [hunk 6]
```diff
@@ -1,19 +1,115 @@
...
+    CHECK(vessel->psychohistory_is_history());
+    barycentre.Add(vessel->psychohistory().last().degrees_of_freedom(),
+                   vessel->mass());
+  }
+  psychohistory_.Append(vessels_.front()->psychohistory().last().time(),
+                  barycentre.Get());
```
Comment (pleroy) [inline ksp_plugin/pile_up.cpp]: Bad indent.

File: ksp_plugin/pile_up.cpp [hunk 7]
```diff
@@ -1,19 +1,115 @@
...
+      prolongation.Append(psychohistory_.last().time(),
+                          psychohistory_.last().degrees_of_freedom());
+      ephemeris.FlowWithAdaptiveStep(
+          &psychohistory_,
+          Ephemeris<Barycentric>::NoIntrinsicAcceleration,
+          t,
+          adaptive_step_parameters,
+          Ephemeris<Barycentric>::unlimited_max_ephemeris_steps);
+      psychohistory_.Append(prolongation.last().time(),
+                      prolongation.last().degrees_of_freedom());
```
Comment (pleroy) [inline ksp_plugin/pile_up.cpp]: Bad indent.

File: ksp_plugin/pile_up.hpp [hunk 8]
```diff
@@ -30,10 +36,41 @@ class PileUp final {
  public:
   explicit PileUp(std::list<not_null<Vessel*>>&& vessels);
 
+  void set_mass_and_intrinsic_force(
```
Comment (pleroy) [inline ksp_plugin/pile_up.hpp]: I would have a preference for two setters, coupling mass and force seems unnecessary.

File: ksp_plugin/pile_up.hpp [hunk 9]
```diff
@@ -30,10 +36,41 @@ class PileUp final {
...
+  void AdvanceTime(
+      Ephemeris<Barycentric>& ephemeris,
+      Instant const& t,
+      Ephemeris<Barycentric>::FixedStepParameters const& fixed_step_parameters,
+      Ephemeris<Barycentric>::AdaptiveStepParameters const&
+          adaptive_step_parameters);
+
  private:
   std::list<not_null<Vessel*>> vessels_;
+  Mass mass_;
```
Comment (pleroy) [inline ksp_plugin/pile_up.hpp]: Should this member and the next be optional, for safety?

File: ksp_plugin/pile_up.hpp [hunk 10]
```diff
@@ -30,10 +36,41 @@ class PileUp final {
...
+  Vector<Force, Barycentric> intrinsic_force_;
+  DiscreteTrajectory<Barycentric> psychohistory_;
+  // True if the last point of the |psychohistory_| should be flowed from;
+  // otherwise, the last point should be removed before flowing the trajectory
+  // (in that case, there is a penultimate point, and it is historical).
+  bool psychohistory_is_history_;
```
Comment (pleroy) [inline ksp_plugin/pile_up.hpp]: As mentioned elsewhere I find this identifier obscure.

File: ksp_plugin/pile_up.hpp [hunk 11]
```diff
@@ -30,10 +36,41 @@ class PileUp final {
...
+  // The |PileUp| is seen as a (currently non-rotating) rigid body; the degrees
+  // of freedom of the vessels in the frame of that can be set, however their
```
Comment (pleroy) [inline ksp_plugin/pile_up.hpp]: This doesn't parse.

File: ksp_plugin/pile_up.hpp [hunk 12]
```diff
@@ -30,10 +36,41 @@ class PileUp final {
...
+  // motion is not integrated; this is simply applied as an offset from the
+  // rigid body motion of the |PileUp|.
+  using RigidPileUp = Frame<serialization::Frame::PluginTag,
+                            serialization::Frame::RIGID_PILE_UP,
+                            /*frame_is_inertial=*/false>;
```
Comment (pleroy) [inline ksp_plugin/pile_up.hpp]: This needs a rigorous description of the frame `RIGID_PILE_UP`.  I assume that its origin is at the centre of mass of the pile-up, but I have no idea about its orientation.

File: ksp_plugin/pile_up.cpp [hunk 13]
```diff
@@ -1,19 +1,115 @@
...
 }
 
+void PileUp::AdvanceTime(
+    Ephemeris<Barycentric>& ephemeris,
+    Instant const& t,
+    Ephemeris<Barycentric>::FixedStepParameters const& fixed_step_parameters,
+    Ephemeris<Barycentric>::AdaptiveStepParameters const&
+        adaptive_step_parameters) {
+  if (!psychohistory_is_history_) {
+    auto const penultimate = --psychohistory_.last();
```
Comment (pleroy) [inline ksp_plugin/pile_up.cpp]: Are we sure that there is a point before `last`?  I'd rather `CHECK`.

File: ksp_plugin/pile_up.cpp [hunk 14]
```diff
@@ -1,19 +1,115 @@
...
+  if (intrinsic_force_ == Vector<Force, Barycentric>{}) {
+    ephemeris.FlowWithFixedStep(
+        {&psychohistory_},
+        Ephemeris<Barycentric>::NoIntrinsicAccelerations,
+        t,
+        fixed_step_parameters);
+    if (psychohistory_.last().time() < t) {
+      // TODO(egg): this is clumsy, we need an option for FlowWithAdaptiveStep
+      // to only add the last point.  Amusingly, this is the unwanted behaviour
+      // of FlowWithFixedStep (#228).
```
Comment (pleroy) [inline ksp_plugin/pile_up.cpp]: This could go in the parameters.  Create an issue?

File: ksp_plugin/interface_vessel.cpp (hunk 1 above)
Comment (eggrobin) [inline ksp_plugin/interface_vessel.cpp]: Done.

File: ksp_plugin/pile_up.cpp (hunk 2 above)
Comment (eggrobin) [inline ksp_plugin/pile_up.cpp]: Done.

File: ksp_plugin/pile_up.cpp (hunk 6 above)
Comment (eggrobin) [inline ksp_plugin/pile_up.cpp]: Done.

File: ksp_plugin/pile_up.cpp (hunk 13 above)
Comment (eggrobin) [inline ksp_plugin/pile_up.cpp]: Adding a CHECK.

File: ksp_plugin/pile_up.cpp (hunk 14 above)
Comment (eggrobin) [inline ksp_plugin/pile_up.cpp]: #1200.

File: ksp_plugin/pile_up.cpp (hunk 7 above)
Comment (eggrobin) [inline ksp_plugin/pile_up.cpp]: Done.

File: ksp_plugin/vessel.cpp (hunk 5 above)
Comment (eggrobin) [inline ksp_plugin/vessel.cpp]: No, the first point has to be authoritative.

File: ksp_plugin/vessel.cpp (hunk 3 above)
Comment (eggrobin) [inline ksp_plugin/vessel.cpp]: OBE.

File: ksp_plugin/vessel.hpp (hunk 4 above)
Comment (eggrobin) [inline ksp_plugin/vessel.hpp]: Done.

File: ksp_plugin/pile_up.hpp (hunk 8 above)
Comment (eggrobin) [inline ksp_plugin/pile_up.hpp]: Done.

File: ksp_plugin/pile_up.hpp (hunk 9 above)
Comment (eggrobin) [inline ksp_plugin/pile_up.hpp]: Initializing them in the constructor instead.

File: ksp_plugin/pile_up.hpp (hunk 10 above)
Comment (eggrobin) [inline ksp_plugin/pile_up.hpp]: Done.

File: ksp_plugin/pile_up.hpp (hunk 11 above)
Comment (eggrobin) [inline ksp_plugin/pile_up.hpp]: Done.

File: ksp_plugin/pile_up.hpp (hunk 12 above)
Comment (eggrobin) [inline ksp_plugin/pile_up.hpp]: Done.

File: ksp_plugin/pile_up.hpp [hunk 15]
```diff
@@ -61,11 +61,14 @@ class PileUp final {
...
+  bool psychohistory_is_authoritative_;
   // The |PileUp| is seen as a (currently non-rotating) rigid body; the degrees
-  // of freedom of the vessels in the frame of that can be set, however their
-  // motion is not integrated; this is simply applied as an offset from the
-  // rigid body motion of the |PileUp|.
+  // of freedom of the vessels in the frame of that body can be set, however
+  // their motion is not integrated; this is simply applied as an offset from
+  // the rigid body motion of the |PileUp|.
+  // The origin of |RigidPileUp| is the centre of mass of the pile up.
+  // its axes are those of Barycentric for now; eventually we will probably want
```
Comment (pleroy) [inline ksp_plugin/pile_up.hpp]: Its

**Pull request merged**